### PR TITLE
Added FreeCAD script to generate a coil with 2 layers

### PR DESCRIPTION
**Pull request by gitansh95**
- FreeCAD script to generate a coil with 2 layers
 - Input parameters like radius of wire, pitch of coil, length of coil, and gap (for a non-intersecting sweep) can be changed from script

### CODE REVIEW

**Comment by mchandra** [comment]
Can you post the image of the coil generated by the script here?

**Comment by gitansh95** [comment]
The images below are of a coil generated using this script with the following parameters :
`wire_radius = 0.10`
`core_height = 10.00`
`core_radius = 5.00`
`gap             = 0.01`

- Axonometric View
axonometric_view

- Front View
front_view

- Top View
top_view

**Comment by mchandra** [comment]
1. What does `gap` correspond to in the above figure?
2. Is this the full setup? Aren't there supposed to be two coils along with a support structure?

**Comment by gitansh95** [comment]
- `gap` is a parameter that I have given so that the two coils do not touch each other and do not create an intersecting sweep. The gap is not visible in the above figures.
- In real world, the wire that will be used to make the coil will consist of a metal wire, coated with an electrically insulating enamel. This gap can also physically model the thickness of that enamel coating.
- I have created the coil as a separate part. I will create the core and the setup as two more separate parts. We will then assemble all these parts in FreeCAD assembly. This will give us the advantage of making changes to one part without necessarily affecting others within the same assembly.
- Also, the 2D drawings for the core need to be separate from the 2 drawings for the coil, the setup, etc, because they may be given to different manufacturers (in-house/external) depending upon the material being used and the complexity of the job.
-  One more advantage that I can think of is that we can have two coils with completely different parameters in this setup, and the only change required would be to tweak the parameters given in the python script.

**Comment by mchandra** [comment]
Sounds good. Is the final assembly of the individual parts (two coils + support structure) also automated, i.e., through a script?

**Comment by gitansh95** [comment]
I have yet to start working on the assembly. But I will do that through a python script too, so that we can easliy control any changes we want to make.

**Comment by mchandra** [comment]
Excellent

**Comment by mchandra** [comment]
1. Post the final image of the experiment.
2. Put up a README detailing how the scripts have to be run in order for the reviewer to check.

**Comment by gitansh95** [comment]
The Python scripts to generate all the models in the Electromagnetic Coupling experiment have been added to the Pull request. The individual part models can be generated using the python scripts given in the code. But the assembly of those parts has to be done manually at the moment. Automating the assembly is in progress.

I have added a README in `Lab-Management/freecad_models` that describes the procedure to build FreeCAD from source, and then run any of the Python scripts in `Lab-Management/freecad_models` in FreeCAD. The parameters for the experiment can be tweaked from the file `experiment_params.py` in the same directory.

The images for the final (manual) assembly of the parts generated using the Python scripts is given below.

- Axonometric View
isometric

- Top View
top


- Front View
front

- Side View
side


The following parameters were take for the above images,
`## Core Parameters`
`core_radius      = 10.0 #mm`
`core_height      = 40.0 #mm`

`# Core-Coil Winding Machine Mount Parameters`
`top_height        = 5.00 #mm`
`base_height     = 15.0 #mm`
`base_cut          = 10.0 #mm`
`coil_thickness   = 5.00 #mm`

`# Core-Experimental Setup Mount Parameters`
`hole_diameter    = 8.50 #mm`
`thread_pitch       = 1.50 #mm`
`chamfer_length  = 4.00 #mm`

`#####################`
`## Coil Parameters`
`wire_radius      = 0.711/2                  # SWG 27 -> wire_radius = 0.711/2`
`gap                  = 0.00                       #mm`
`coil_height       = core_height           # Coil height will be modified slightly to accomodate an integral number of turns`
`coil_radius      = core_radius + wire_radius + gap # This is the radius of the inner most helix of the coil`

`#####################`
`## Cover Parameters`
`top_flange_width = 10.0 #mm`
`side_margin      = 25.0 #mm`
`length_margin    = 25.0 #mm`

`sheet_thickness  = 2.00 #mm`
`bend_radius        = 1.00 #mm`
`screw_diameter  = 2.00 #mm`
`screw_length      = 10.0 #mm`

`#Coaxial connector parameters`
`coax_side_distance      = 5.00 #mm`
`coax_bottom_distance = 55.0 #mm`
`coax_base_width          = 5.75 #mm`
`coax_radius                   = 4.90 #mm`

**Comment by shyams2** [comment]
1. README for the FreeCAD models needs to be merged into the README of the root directory(or mentioned that the instructions may be found under the specific directory)

2. SUGGESTION: Might be a better idea to provide a link to the page that contains the build instructions as provided by the FreeCAD developers, since not everyone might be running Ubuntu(@mchandra, thoughts?)

3. It'll be great if you could provide a build guide where dependencies need not be installed using the package manager, and could be built locally. That way, it'll be possible to run your scripts on a computer without having root access.

4. Building dependencies gives the following errors:
```bash
E: Unable to locate package liboce-foundation8
E: Unable to locate package liboce-modeling8
E: Unable to locate package liboce-ocaf-lite8
E: Unable to locate package liboce-ocaf8
E: Unable to locate package liboce-visualization8
```
Is this a result of Ubuntu 16.04 which is fixed by the instructions provided later?

5. It is not clear to what platform the build is carried out on when the approximate time for build is mentioned. 

6. In the README, the following instructions have been mentioned for the process of building:

Navigate to the folder where you cloned freecad : cd free-cad-code
Now type make

However, this results in the error:
`make: *** No targets specified and no makefile found.  Stop.`

This needs to be changed to:
```bash
mkdir freecad-build
cd freecad-build
cmake ..
make 
```

7. I found a lot of difficulties with the build. In addition to the mentioned instructions, I had to run the following commands:
```bash
sudo apt install blt fonts-lyx gstreamer1.0-plugins-good gstreamer1.0-pulseaudio gstreamer1.0-x libaa1 libboost-program-options1.58.0 libboost-regex1.58.0 libboost-signals1.58.0 libcoin80v5 libdv4 libglade2-0 libgstreamer-plugins-good1.0-0 libjs-jquery-ui liboce-foundation10 liboce-modeling10 liboce-ocaf-lite10 liboce-ocaf10 liboce-visualization10 libphonon4 libpulse-mainloop-glib0 libpyside1.2 libqt4-designer libqt4-help libqt4-opengl libqt4-scripttools libqt4-svg libqt4-test libqtassistantclient4 libqtwebkit4 libshiboken1.2v5 libshout3 libsoqt4-20 libspnav0 libtag1v5 libtag1v5-vanilla libv4l-0 libv4lconvert0 libzipios++0v5 phonon phonon-backend-gstreamer phonon-backend-gstreamer-common pyside-tools python-bs4 python-cairo python-collada python-cycler python-dateutil *boost* python-glade2 python-gobject-2 python-gtk2 python-html5lib python-imaging python-lxml python-matplotlib python-matplotlib-data python-numpy python-pil python-pivy python-ply python-pyparsing python-pyside python-pyside.phonon python-pyside.qtcore python-pyside.qtdeclarative python-pyside.qtgui python-pyside.qthelp python-pyside.qtnetwork python-pyside.qtopengl python-pyside.qtscript python-pyside.qtsql python-pyside.qtsvg python-pyside.qttest python-pyside.qtuitools python-pyside.qtwebkit
python-pyside.qtxml python-qt4 python-qt4-gl python-sip python-tk python-tz tk8.6-blt2.5 ttf-bitstream-vera
oce-draw liboce* libqt4* libcoin* libshiboken* libpyside*
```

8. Generating the models often resulted in FreeCAD hanging and crashing. However, I believe this is an 
   issue caused due to my laptop. But after waiting for sufficiently long enough, I was able to build the 
   following parts:
   - core
   - coil_winding
   - mount

9. I got the following error while running `execfile('cover_base.py')
```python
>>> execfile('cover_base.py')
Traceback (most recent call last):
  File "<input>", line 1, in <module>
  File "cover_base.py", line 12, in <module>
    import SheetMetalCmd
ImportError: No module named SheetMetalCmd
```
A similar error was obtained also while executing cover_top.py. Do I need to get some additional packages?

- I have the following issues with the code:

    - In most of the files, you seem to have used `from experiment_params import *`. This is a bad 
      practice, and is recommended against since it isn't explicitly clear as to what has been imported.

    - What is the purpose of this line?:`sys.path.insert(0, 
      "${HOME}/Downloads/susceptibility_experiment/code")`

    - Though the language seems to be explanatory of what is happening most of the times, could you 
       possibly add more comments to make every step clear. For instance in core.py:
```python
App.newDocument("core")
App.setActiveDocument("core")
App.ActiveDocument=App.getDocument("core")
Gui.ActiveDocument=Gui.getDocument("core")
# I'm not able to get what App & Gui refer to here. From my understanding, App is a class whose methods are used to perform all the steps in constructing the object, while Gui's methods are used to change appearances in the GUI/Send messages?

# What is happening here?:
Gui.SendMsgToActiveView("ViewFit")
Gui.activeDocument().activeView().viewAxometric()

#Cylinder 1
App.ActiveDocument.addObject("Part::Cylinder","Cylinder")
App.ActiveDocument.ActiveObject.Label = "Cylinder"
App.ActiveDocument.recompute() # What is recompute?
Gui.SendMsgToActiveView("ViewFit") # Same command as above

FreeCAD.getDocument("core").getObject("Cylinder").Radius = core_radius
FreeCAD.getDocument("core").getObject("Cylinder").Height = core_height

# Again:
App.ActiveDocument.recompute()
Gui.SendMsgToActiveView("ViewFit")

# What are the numbers being used in the arguments for Vector - (x, y, z)?
# It'll be helpful if mentioned explicitly.
FreeCAD.getDocument("core").getObject("Cylinder").Placement = App.Placement(App.Vector(0,0,base_height),App.Rotation(App.Vector(0,0,1),0))
# Not able to understand the purpose of Rotation
..
....
..
App.activeDocument().addObject("Part::MultiFuse","Fusion")
App.activeDocument().Fusion.Shapes = [App.activeDocument().Cylinder,App.activeDocument().Cylinder002,App.activeDocument().Cylinder001,]
# Why is visibility toggled to false?
Gui.activeDocument().Cylinder.Visibility=False
Gui.activeDocument().Cylinder002.Visibility=False
Gui.activeDocument().Cylinder001.Visibility=False
Gui.ActiveDocument.Fusion.ShapeColor=Gui.ActiveDocument.Cylinder002.ShapeColor
Gui.ActiveDocument.Fusion.DisplayMode=Gui.ActiveDocument.Cylinder002.DisplayMode
App.ActiveDocument.recompute()

# What is this?
Gui.activateWorkbench("PartWorkbench")

App.ActiveDocument.addObject("Part::Box","Box")
App.ActiveDocument.ActiveObject.Label = "Cube"

FreeCAD.getDocument("core").getObject("Box").Length = core_radius + coil_thickness
FreeCAD.getDocument("core").getObject("Box").Width = 2*(core_radius + coil_thickness)
FreeCAD.getDocument("core").getObject("Box").Height = base_cut

# I understand this may be pretty complicated. But is there any way to possibly explain, why you have used those parameters the way you've used them:
FreeCAD.getDocument("core").getObject("Box").Placement = \
    App.Placement(App.Vector((core_radius+coil_thickness)/2+chamfer_length,-(core_radius+coil_thickness),0),App.Rotation(App.Vector(0,0,1),0))

#cut the part

# Can't understand. What is being cut? The fused cylinders? Along what?
App.activeDocument().addObject("Part::Cut","Cut")
App.activeDocument().Cut.Base = App.activeDocument().Fusion
App.activeDocument().Cut.Tool = App.activeDocument().Box
Gui.activeDocument().hide("Fusion")
Gui.activeDocument().hide("Box")
Gui.ActiveDocument.Cut.ShapeColor=Gui.ActiveDocument.Fusion.ShapeColor
Gui.ActiveDocument.Cut.DisplayMode=Gui.ActiveDocument.Fusion.DisplayMode

App.ActiveDocument.recompute()
Gui.SendMsgToActiveView("ViewFit")
```

**Comment by gitansh95** [comment]
1. As of now, the FreeCADCAD code to generate models is not related in any way to the Lab-Management's code that enables users to write and perform experiments, but we do plan to integrate both of these. And I will merge the README files as soon as that is done.

2. Yeah, I think that is a good idea. I thought I would document everything I did to get my build up and running. We'll take a third opinion from @mchandra .

3. I think the general purpose libraries can be installed from the package manager. Although FreeCAD can also be installed from the package manager, we have chosen to work with the latest release (from github). This gives us access to more and better refined features, and fewer bugs.

4. In the terminal, type `liboce-` and double press tab key to see what version of `lib-oce` files you have. In my case I have `liboce-foundation10`, `liboce-ocaf10`, `lib-modelling10` etc. Install the available libs corresponding to the ones mentioned in the README.

5. This will depend on how point number 2 is solved. We may remove that line from the README.

6. That is a mistake on my part. It needs to be changed.

7. I see that you have installed `liboce*` in this step. That must have solved the point number 4.

8. Yes, these parts are computationally intensive because they contain helixes with a small pitch.  The helixes in the parts you have mentioned causes my laptop to freeze for a while, and when I wait long enough, the part gets generated.

9. A mistake on my part. You need to install the following 3rd party workbenches in the folder where FreeCAD is installed, in order for all features of the code to work.
Navigate to the FreeCAD installation directory:
`cd path/to/FreeCAD/Mod/`
Clone the following repositories :
- Sheet Metal : `git clone https://github.com/shaise/FreeCAD_SheetMetal.git`
- Assembly2 :  `git clone https://github.com/hamish2014/FreeCAD_assembly2.git`
The next time you run FreeCAD, in the terminal it should print `Sheet metal workbench loaded` Also, in the dropdown menu for workbenches in the FreeCAD GUI, you should be able to see `Assembly2` and `Sheet Metal`.

10. 
```
App.newDocument("core")
App.setActiveDocument("core")
App.ActiveDocument=App.getDocument("core")
Gui.ActiveDocument=Gui.getDocument("core")
# FreeCAD is split into App and GUI modules. App gives you access to defining the properties of
#shapes, and using GUI you can change the way you want to represent the shape.

#I would like to point you to this link for more info : https://www.freecadweb.org/wiki/FreeCAD_Scripting_Basics#Built-in_modules

Gui.SendMsgToActiveView("ViewFit") 
#Tells the GUI to fit the object inside the screen. This statement is not necessary, but I have added it
#each time I create a new geometry so that the user is always seeing the complete representation of
#the CAD model and which fits inside the screen.

Gui.activeDocument().activeView().viewAxometric()
#Sets the view to axonometric (wiki link https://en.wikipedia.org/wiki/Axonometric_projection)

#Cylinder 1
App.ActiveDocument.addObject("Part::Cylinder","Cylinder")
App.ActiveDocument.ActiveObject.Label = "Cylinder"

App.ActiveDocument.recompute()
# Recomputes the document - and updates all properties that have been changed inside the document.

Gui.SendMsgToActiveView("ViewFit") # Same command as above

FreeCAD.getDocument("core").getObject("Cylinder").Radius = core_radius
FreeCAD.getDocument("core").getObject("Cylinder").Height = core_height

# Again:
App.ActiveDocument.recompute()
Gui.SendMsgToActiveView("ViewFit")

# What are the numbers being used in the arguments for Vector - (x, y, z)?
# It'll be helpful if mentioned explicitly.
FreeCAD.getDocument("core").getObject("Cylinder").Placement = App.Placement(App.Vector(0,0,base_height),App.Rotation(App.Vector(0,0,1),0))
# Placement property consists of 3 things
# The position of the base (vector)
# The axis about which the object is to be rotated (In the above code it is the z axis (0, 0, 1) because x
# and y are 0, and z is 1. )
# Rotation Angle - this rotates the object by the specified angle along the axis specified above 
# (In the above line the angle is 0 : see the final argument of App.Placement()).
..
....
..
App.activeDocument().addObject("Part::MultiFuse","Fusion")
App.activeDocument().Fusion.Shapes = [App.activeDocument().Cylinder,App.activeDocument().Cylinder002,App.activeDocument().Cylinder001,]
# Why is visibility toggled to false?
# Cylinder, Cylinder001, and Cylinder002 were fused to form a new part - Fusion
# The old subparts are made invisible, while only the new complete part, Fusion, is visible.
Gui.activeDocument().Cylinder.Visibility=False
Gui.activeDocument().Cylinder002.Visibility=False
Gui.activeDocument().Cylinder001.Visibility=False
Gui.ActiveDocument.Fusion.ShapeColor=Gui.ActiveDocument.Cylinder002.ShapeColor
Gui.ActiveDocument.Fusion.DisplayMode=Gui.ActiveDocument.Cylinder002.DisplayMode
App.ActiveDocument.recompute()

# What is this?
# Workbenches contain all the functionality in FreeCAD. Depending on what you want to do, you need
# to use a different workbench. Here, to create a Part::Box, we need the part workbench.
# Since we are using only the code to generate the parts, I guess removing the line below may not 
# affect the process.
Gui.activateWorkbench("PartWorkbench")

App.ActiveDocument.addObject("Part::Box","Box")
App.ActiveDocument.ActiveObject.Label = "Cube"

FreeCAD.getDocument("core").getObject("Box").Length = core_radius + coil_thickness
FreeCAD.getDocument("core").getObject("Box").Width = 2*(core_radius + coil_thickness)
FreeCAD.getDocument("core").getObject("Box").Height = base_cut

# I understand this may be pretty complicated. But is there any way to possibly explain, why you have used those parameters the way you've used them:
FreeCAD.getDocument("core").getObject("Box").Placement = \
    App.Placement(App.Vector((core_radius+coil_thickness)/2+chamfer_length,-(core_radius+coil_thickness),0),App.Rotation(App.Vector(0,0,1),0))

#cut the part

# Can't understand. What is being cut? The fused cylinders? Along what?
# The base for the cut is 'Fusion' (cylinders fused above). The Box created above is the cut tool.
# Basically a subtraction operation is being performed which is equivalent to Cut = Fusion - Box
App.activeDocument().addObject("Part::Cut","Cut")
App.activeDocument().Cut.Base = App.activeDocument().Fusion
App.activeDocument().Cut.Tool = App.activeDocument().Box
Gui.activeDocument().hide("Fusion")
Gui.activeDocument().hide("Box")
Gui.ActiveDocument.Cut.ShapeColor=Gui.ActiveDocument.Fusion.ShapeColor
Gui.ActiveDocument.Cut.DisplayMode=Gui.ActiveDocument.Fusion.DisplayMode

App.ActiveDocument.recompute()
Gui.SendMsgToActiveView("ViewFit")
```

**Review by shyams2 — CHANGES_REQUESTED** [review]
- It'll be nice if the properties a user needs to keep in mind while defining the parameters are mentioned. Since there are certain cases where the build will fail maybe because of overlap etc..

- Currently, `assembly.py` fails. This is due to `XPL-XLIA.fcstd` not being available. I changed this to `Lockin_base.STEP`. However, even with this change the build fails giving the error `A part can only be imported from a FreeCAD document with exactly one visible part. Aborting operation`

- I've made a [PR](https://github.com/gitansh95/Lab-Management/pull/2) containing comments of my understanding of the code and questioned ambiguous sections. I've added a lot of comments in `core.py` with the elements unique to other files being commented upon. Please make sure that the commenting is also followed in all the other files to make the steps clear.

- Please change the formatting used in the code. Some lines go uptil the 141st column

**Review by shyams2 — APPROVED** [review]
